### PR TITLE
Vote

### PR DESCRIPTION
**Pull request by GTZL1**
Start Vote implementation

### CODE REVIEW

File: src/main/java/ch/heigvd/pro/b04/answers/Answer.java [hunk 1]
```diff
@@ -1,7 +1,14 @@
 package ch.heigvd.pro.b04.answers;
 
+import ch.heigvd.pro.b04.participants.Participant;
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/answers/Answer.java]: Useless import

File: src/main/java/ch/heigvd/pro/b04/votes/Vote.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import ch.heigvd.pro.b04.answers.Answer;
+import ch.heigvd.pro.b04.participants.Participant;
+import javax.persistence.EmbeddedId;
+import javax.persistence.Entity;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Entity
+@NoArgsConstructor
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/Vote.java:16]: We probably want a Builder as well, to be able to test it correctly and easily

File: src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+@Data
+@Embeddable
+public class VoteIdentifier implements Serializable {
+
+  @Column
+  private Timestamp timeVote;
+
+  @ManyToOne
+  @PrimaryKeyJoinColumn
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java:30]: I'm not sure I understand what `@PrimaryKeyJoinColumn` does ?

**Review by glsubri — CHANGES_REQUESTED** [review]
Also, the PR is missing tests

File: src/main/java/ch/heigvd/pro/b04/answers/Answer.java (hunk 1 above)
Comment (GTZL1) [reply src/main/java/ch/heigvd/pro/b04/answers/Answer.java]: If there's only this.... it's a detail seriously

File: src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java (hunk 3 above)
Comment (GTZL1) [reply src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java:30]: Me neither but I know it's necessarry

File: src/main/java/ch/heigvd/pro/b04/votes/Vote.java (hunk 2 above)
Comment (GTZL1) [reply src/main/java/ch/heigvd/pro/b04/votes/Vote.java:16]: ok...

**Comment by GTZL1** [comment]
It now fixes #111

File: src/main/java/ch/heigvd/pro/b04/answers/Answer.java [hunk 4]
```diff
@@ -28,4 +37,18 @@ public Answer(long id, String title, String description) {
     this.text = title;
     this.description = description;
   }
+
+  /**
+   * add a {@link Vote} to the current {@link Answer}.
+   *
+   * @param newVote vote to add
+   */
+  public void addVote(Vote newVote) {
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/answers/Answer.java]: I think we can remove it ?

File: src/main/java/ch/heigvd/pro/b04/participants/Participant.java [hunk 5]
```diff
@@ -35,4 +43,25 @@ public static Long getNewIdentifier(ParticipantRepository repository) {
   @Column(unique = true)
   @Getter
   private String token;
+
+  @OneToMany(mappedBy = "idVote.idxParticipant", cascade = CascadeType.ALL)
+  private Set<Vote> voteSet;
+
+  public Participant(long id) {
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/participants/Participant.java]: Not needed

File: src/main/java/ch/heigvd/pro/b04/participants/Participant.java [hunk 6]
```diff
@@ -35,4 +43,25 @@ public static Long getNewIdentifier(ParticipantRepository repository) {
...
+    idParticipant = new ParticipantIdentifier(id);
+  }
+
+  /**
+   * add a {@link Vote} to the current {@link Participant}.
+   *
+   * @param newVote vote to add
+   */
+  public void addVote(Vote newVote) {
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/participants/Participant.java]: We do not need this method either

File: src/main/java/ch/heigvd/pro/b04/votes/Vote.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+  @Getter
+  @EmbeddedId
+  private VoteIdentifier idVote;
+
+  @Getter
+  @Setter
+  private boolean answerChecked;
+
+  public Vote(Participant participant, Answer answer) {
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/Vote.java]: Not needed

File: src/main/java/ch/heigvd/pro/b04/votes/VoteController.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+   * @param token    voter's token
+   * @param chekced  if this answer is choosed or not
+   * @param idAnswer answer concerned by vote
+   * @throws ResourceNotFoundException    if Participant or Answer doesn't exist
+   * @throws SessionNotAvailableException if Session is in state Closed
+   */
+  @PutMapping(value = "/mod/{idModerator}/poll/{idPoll}"
+      + "/question/{idQuestion}/answer/{idAnswer}/vote")
+  public void newVote(@RequestParam(name = "token") String token,
+      @RequestBody boolean chekced,
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/VoteController.java]: typo -> "checked"

File: src/main/java/ch/heigvd/pro/b04/votes/VoteController.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    Optional<Participant> voter = participantRepository.findByToken(token);
+    Optional<Answer> answerChanged = answerRepository.findById(idAnswer);
+
+    if (voter.isEmpty() || answerChanged.isEmpty()) {
+      throw new ResourceNotFoundException();
+    }
+
+    Optional<Session> session = sessionRepository.findById(
+        voter.get().getIdParticipant().getIdxSession().getIdSession());
+    if (session.isEmpty() || session.get().getState() == State.CLOSED) {
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/VoteController.java]: `session.isEmpty()` should throw a `NotExisting`

File: src/main/java/ch/heigvd/pro/b04/votes/VoteController.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+  private final ParticipantRepository participantRepository;
+  private final VoteRepository voteRepository;
+  private final AnswerRepository answerRepository;
+  private final SessionRepository sessionRepository;
+
+  /**
+   * Create and store a new {@link Vote}.
+   *
+   * @param token    voter's token
+   * @param chekced  if this answer is choosed or not
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/VoteController.java]: "choosed" -> "chosen"

File: src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+import javax.persistence.PrimaryKeyJoinColumn;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Builder
+@NoArgsConstructor
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java:20]: Missing `@AllArgsConstructor`

File: src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+  @PrimaryKeyJoinColumn
+  private Answer idxAnswer;
+
+  /**
+   * VoteIdentifier constructor.
+   *
+   * @param participant idxParticipant
+   * @param answer      idxAnswer
+   */
+  public VoteIdentifier(Participant participant, Answer answer) {
```
Comment (glsubri) [inline src/main/java/ch/heigvd/pro/b04/votes/VoteIdentifier.java]: Not needed

**Review by glsubri — CHANGES_REQUESTED** [review]

**Review by glsubri — APPROVED** [review]

**Pull request merged**